demo-circuit/src/blaze_csw/constraints/mod.rs

```diff
@@ -0,0 +1,1682 @@
+                            - ft_data.sc_txs_com_hashes.len()
+                    ])
+                }
+                Ok(CswProverData {
+                    sys_data,
+                    last_wcert: WithdrawalCertificateData::get_phantom(num_custom_fields),
+                    utxo_data: CswUtxoProverData::get_phantom(),
+                    ft_data,
+                })
+            }
+            // Attempt to withdraw a sc utxo without having specified a last_wcert
+            (None, Some(_), _) => Err(Error::from(
+                "Attempt to withdraw SC Utxo without specifying last WCert",
+            )),
+            // Attempt to withdraw both a sc utxo and a ft
+            (_, Some(_), Some(_)) => Err(Error::from(
+                "Cannot create a CSW proof for retrieving both a SC UTXO and a FT",
+            )),
+            // Any other combination is not admissable
+            _ => Err(Error::from("Unexpected inputs combination")),
+        }?;
+
+        Ok(Self {
+            range_size,
+            num_custom_fields,
+            sidechain_id,
+            csw_data,
+            constant,
+            csw_sys_data_hash,
+        })
+    }
+
+    // For testing, if useful
+    pub fn from_prover_data(
+        sidechain_id: FieldElement,
+        constant: Option<FieldElement>,
+        csw_data: CswProverData,
+        range_size: u32,
+        num_custom_fields: u32,
+    ) -> Result<Self, Error> {
+        let csw_sys_data_hash = Self::compute_csw_sys_data_hash(&csw_data.sys_data, sidechain_id)?;
+
+        Ok(CeasedSidechainWithdrawalCircuit {
+            sidechain_id,
+            csw_data,
+            range_size,
+            num_custom_fields,
+            constant,
+            csw_sys_data_hash,
+        })
+    }
+
+    pub fn get_instance_for_setup(
+        range_size: u32,
+        num_custom_fields: u32,
+        is_constant_present: bool,
+    ) -> Self {
+        Self {
+            range_size,
+            num_custom_fields,
+            sidechain_id: FieldElement::default(),
+            csw_data: CswProverData::get_default(range_size, num_custom_fields),
+            constant: if is_constant_present {
+                Some(FieldElement::default())
+            } else {
+                None
+            },
+            csw_sys_data_hash: FieldElement::default(),
+        }
+    }
+}
+
+impl ConstraintSynthesizer<FieldElement> for CeasedSidechainWithdrawalCircuit {
+    fn generate_constraints<CS: ConstraintSystemAbstract<FieldElement>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+        // Alloc sidechain id
+        let sidechain_id_g =
+            FieldElementGadget::alloc(cs.ns(|| "alloc sidechain_id_g"), || Ok(&self.sidechain_id))?;
+
+        // Alloc all witness data
+        let csw_data_g =
+            CswProverDataGadget::alloc(cs.ns(|| "alloc csw data"), || Ok(&self.csw_data))?;
+
+        // Decide whether to enforce utxo or ft withdrawal
```

Comment:
When reading it the first time, the independent choice of the the `should_enforc_utxo/ft_withdrawal_g` made me wonder. 
Maybe it is worth to mention in a comment that it is as good as impossible to satisfy both a utxo and a ft withdrawal within this circuit.

Reply:
Added a comment about this